src/core/api.rs

```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
+        encaps(
+            &mut *self.rng.lock().expect("Mutex lock failed!"),
+            mpk,
+            &policy.generate_point_coordinates(ap.clone())?,
+        )
+    }
+    /// Attempts opening the given encapsulation using the given
+    /// user secret key.
+    ///
+    /// Returns the encapsulated symmetric key if the user key holds
+    /// the correct rights.
+    fn decaps(
+        &self,
+        usk: &UserSecretKey,
+        encapsulation: Encapsulation,
+    ) -> Result<Option<Secret<SEED_LENGTH>>, Error> {
+        decaps(usk, &encapsulation)
+    }
+}
+
+pub trait CovercryptPKE<
+    Aead,
+    const KEY_LENGTH: usize,
+    const NONCE_LENGTH: usize,
+    const MAC_LENGTH: usize,
+>
+{
+    /// Encrypts the given plaintext using Covercrypt and the given DEM.
+    ///
+    /// Creates a Covercrypt encapsulation of a LENGTH-byte key, and use this
+    /// key with the given authentication data to produce a DEM ciphertext of
+    /// the plaintext.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encrypt(
+        &self,
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+        ad: Option<&[u8]>,
+        plaintext: &[u8],
+    ) -> Result<(Encapsulation, Vec<u8>), Error>;
+
+    /// Attempts decrypting the given ciphertext using the Covercrypt KEM and the DEM.
+    ///
+    /// Attempts opening the Covercrypt encapsulation. If it succeeds, decrypts
+    /// the ciphertext using the DEM with the encapsulated key and the given
+    /// authentication data. Returns ‘None‘ otherwise.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn decrypt(
+        &self,
+        usk: &UserSecretKey,
+        ad: Option<&[u8]>,
+        cyphertext: &[u8],
+        enc: &Encapsulation,
+    ) -> Result<Option<Vec<u8>>, Error>;
+}
+
+impl<
+        const KEY_LENGTH: usize,
+        const NONCE_LENGTH: usize,
+        const MAC_LENGTH: usize,
+        E: AE<KEY_LENGTH, NONCE_LENGTH, MAC_LENGTH>,
+    > CovercryptPKE<E, KEY_LENGTH, NONCE_LENGTH, MAC_LENGTH> for Covercrypt
+{
+    /// Encrypts the given plaintext using Covercrypt and the given DEM.
+    ///
+    /// Creates a Covercrypt encapsulation of a LENGTH-byte key, and use this
+    /// key with the given authentication data to produce a DEM ciphertext of
+    /// the plaintext.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encrypt(
+        &self,
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+        ad: Option<&[u8]>,
+        plaintext: &[u8],
+    ) -> Result<(Encapsulation, Vec<u8>), Error> {
+        let (seed, enc) = CovercryptKEM::encaps(self, mpk, policy, ap)?;
+        if seed.length() < KEY_LENGTH {
+            return Err(Error::KeyError(
+                String::from_str("SEED_LENGTH >= KEY_LENGTH").unwrap(),
+            ));
+        }
```

Comment:
- error message is wrong or at least misleading;
- just rehash the seed instead